### PR TITLE
Configurable request matching

### PR DESCRIPTION
**Pull request by dmcrodrigues**
This PR contains the infrastructure to support #3.

Attributes supported:
- [x] Method
- [x] URL
- [x] Path
- [x] Query
- [x] Headers
- [x] Body
- [x] Custom

### CODE REVIEW

**Comment by RuiAAPeres** [comment]
Should we lock this piece of work @dmcrodrigues ?

**Comment by dmcrodrigues** [comment]
I think we can :lock: this one and start reviewing it :eyeglasses:

File: Vinyl/RequestMatcherRegistry.swift [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+struct PathRequestMatcher: RequestMatcher {
+    func match(aRequest: Request, anotherRequest: Request) -> Bool {
+        return aRequest.URL?.path == anotherRequest.URL?.path
+    }
+}
+
+struct QueryRequestMatcher: RequestMatcher {
+    func match(aRequest: Request, anotherRequest: Request) -> Bool {
+        
+        let queryItems: Request -> [NSURLQueryItem]  = { request in
```
Comment (dmcrodrigues) [inline Vinyl/RequestMatcherRegistry.swift]: Double space :oncoming_police_car: 
` [NSURLQueryItem]  =`

File: Vinyl/RequestMatcherRegistry.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            let components = NSURLComponents.init(string: request.URL?.absoluteString ?? "")
```
Comment (dmcrodrigues) [inline Vinyl/RequestMatcherRegistry.swift]: @RuiAAPeres any strong reason to explicit `init`?

File: Vinyl/Player.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    let vinyl: Vinyl
+    let trackMatcher: TrackMatcher
+    
+    private func seekTrackForRequest(request: Request) -> Track? {
+        for track in vinyl.tracks {
+            if trackMatcher.matchableTrack(request, track: track) {
+                return track
+            }
+        }
+        return nil
```
Comment (RuiAAPeres) [inline Vinyl/Player.swift]: this is pretty much a filter + first

File: Vinyl/RequestMatcherRegistry.swift [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+protocol RequestMatcher {
+    func match(aRequest: Request, anotherRequest: Request) -> Bool
+}
+
+struct RequestMatcherRegistry {
+    
+    let registeredTypes: [RequestMatcherType]
+
```
Comment (RuiAAPeres) [inline Vinyl/RequestMatcherRegistry.swift]: private registeredTypes?

File: Vinyl/RequestMatcherRegistry.swift [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        registeredTypes = types
+        matchingChain = types.map { RequestMatcherRegistry.matcherForType($0) }
+    }
+    
+    func matchableRequests(aRequest: Request, anotherRequest: Request) -> Bool {
+        for requestMatcher in matchingChain {
+            if !requestMatcher.match(aRequest, anotherRequest: anotherRequest) {
+                return false
+            }
+        }
```
Comment (RuiAAPeres) [inline Vinyl/RequestMatcherRegistry.swift]: this is a map?

File: Vinyl/Track.swift [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        
+        // We're using a helper function because we cannot mutate a NSURLRequest directly
+        request = Request.createWithEncodedRequest(encodedRequest)
+        
+        response = Response(encodedResponse: encodedResponse)
+    }
+}
+
+extension NSURLRequest {
+
```
Comment (RuiAAPeres) [inline Vinyl/Track.swift]: Move this to the folder Extensions/

File: Vinyl/Track.swift [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        if let body = encodedRequest["body"] as? NSData {
+            request.HTTPBody = body
+        }
+        
+        return request
+    }
+}
+
+extension Response {
+
```
Comment (RuiAAPeres) [inline Vinyl/Track.swift]: Same thing for this one (/Extensions)

**Comment by dmcrodrigues** [comment]
Seems good @RuiAAPeres, I think we can merge it.

**Comment by dmcrodrigues** [comment]
:steam_locomotive:

**Comment by RuiAAPeres** [comment]
close #3

**Pull request merged**